File: rust/36.valid-sudoku.rs
/*
 * @lc app=leetcode id=36 lang=rust
 *
 * [36] Valid Sudoku
 */

// @lc code=start
impl Solution {
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        // Solution 1: HashSet
        // use std::collections::HashSet;
        // let mut row_set = HashSet::new();
        // let mut col_set = HashSet::new();
        // let mut box_set = HashSet::new();
        // for row in 0..9 {
        //     for col in 0..9 {
        //         match board[row][col] {
        //             '.' => continue,
        //             cell => {
        //                 if !row_set.insert((row, cell)) {
        //                     return false;
        //                 }
        //                 if !col_set.insert((col, cell)) {
        //                     return false;
        //                 }
        //                 if !box_set.insert((row / 3 * 3 + col / 3, cell)) {
        //                     return false;
        //                 }
        //             }
        //         }
        //     }
        // }

        // Solution 2:
        let mut rows = [0; 9];
        let mut cols = [0; 9];
        let mut boxes = [0; 9];
        for r in 0..9 {
            for c in 0..9 {
                match board[r][c] {
                    '.' => continue,
                    cell => {
                        let b = (r / 3 * 3) + c / 3;
                        let d = 1 << (cell.to_digit(10).unwrap() - 1);
                        if rows[r] & d != 0 || cols[c] & d != 0 || boxes[b] & d != 0 {
                            return false;
                        }
                        rows[r] += d;
                        cols[c] += d;
                        boxes[b] += d;
                    }
                }
            }
        }
        true
    }
}
// @lc code=end

File: javascript/36.valid-sudoku.js
/*
 * @lc app=leetcode id=36 lang=javascript
 *
 * [36] Valid Sudoku
 */

// @lc code=start
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function (board) {
  const row = new Array(9).fill(0);
  const col = new Array(9).fill(0);
  const box = new Array(9).fill(0);

  for (let r = 0; r < 9; r++) {
    for (let c = 0; c < 9; c++) {
      const cell = board[r][c];
      if (cell === ".") continue;
      const b = ~~(r / 3) * 3 + ~~(c / 3);
      const d = 1 << (cell - 1);
      if ((row[r] & d) !== 0 || (col[c] & d) !== 0 || (box[b] & d) !== 0)
        return false;
      row[r] += d;
      col[c] += d;
      box[b] += d;
    }
  }
  return true;
};
// @lc code=end
